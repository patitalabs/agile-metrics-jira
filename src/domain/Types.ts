import {
  PtsCollectorConfig,
  PtsMetricItem,
} from './project-tracking-system/Types';

export interface TeamMetricsRequest {
  shouldUpdateEntries: boolean;
  config: any;
}

export interface CoreMetricsService {
  publish(entries: any, shouldReplaceEntries: boolean);
}

export interface PtsCollectorService {
  fetch(ciCollectorConfig: PtsCollectorConfig): Promise<PtsMetricItem[]>;
}
