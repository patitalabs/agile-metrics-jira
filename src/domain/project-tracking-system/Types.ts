export interface PtsMetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  key: string;
  createdBy: string;
  issueType: string;
  movedForward: number;
  movedBackward: number;
  storyPoints: number;
  assignees: string[];
  tags: string[];
  created: Date;
  finished: Date;
  leadTime: number;
  cycleTime: number;
  commentCount: number;
  jiraProject: string;
  teamName: string;
  estimateHealth: number;
  rawEstimateHealth: number;
  numberOfBugs: number;
}

export class PtsCollectorConfig {
  teamId: number;
  teamName: string;
  workFlowType: 'sprint' | 'kanban';
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  fields?: { [name: string]: string };
  estimateConfig?: { maxTime: number; estimationValues: number[] };

  constructor({
    teamId,
    since,
    until = null,
    workFlowMap,
    fields,
    teamName,
    workFlowType,
  }) {
    this.teamId = teamId;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.workFlowMap = workFlowMap;
    this.fields = fields;
    this.teamName = teamName;
    this.workFlowType = workFlowType ? workFlowType : 'sprint';
  }

  isKanban(): boolean {
    return this.workFlowType === 'kanban';
  }

  isSprint(): boolean {
    return this.workFlowType === 'sprint';
  }
}
