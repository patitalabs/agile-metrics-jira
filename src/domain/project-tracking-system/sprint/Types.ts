export interface TaskStatistics {
  moveBackward: number;
  moveForward: number;
  movedToDev: Date;
}
