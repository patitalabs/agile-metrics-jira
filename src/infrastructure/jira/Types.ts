export class Sprint {
  id: number;
  name: string;
  isoStartDate: Date;
  isoEndDate: Date;
}

export class Task {
  key: string;
  typeName: string;
  statusName: string;
  assignee: string;
  epic: string;
  sprint?: Sprint;
  created: Date;
  projectName: string;
  teamName: string;
  createdBy: string;
  resolutionDate: Date;
  storyPoints: number;
  labels: string[];
  subtasks: Subtask[];
  histories: { [type: string]: HistoryEntry[] };
  numberOfComments: number;
  numberOfBugs: number;
}

export class Subtask {
  key: string;
  statusName: string;
}

export class HistoryItem {
  field: string;
  fromString: string;
  toString: string;
}

export class HistoryEntry {
  category: string;
  created: Date;
  items: HistoryItem[];
}

export interface JiraConfig {
  teamId: number;
  teamName: string;
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  workFlowType: 'sprint' | 'kanban';
  fields?: { [name: string]: string };
}

export interface IssueDetails {
  subtasks: Subtask[];
  histories: { [type: string]: HistoryEntry[] };
  labels: string[];
  created: Date;
  createdBy: string;
  resolutionDate: Date;
  projectName: string;
  teamName: string;
  numberOfComments: number;
  numberOfBugs: number;
}

export interface JiraService {
  completedSprintsSince(
    teamId: number,
    since: Date,
    until: Date
  ): Promise<Sprint[]>;

  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]>;

  completedKanbanIssuesSince(jiraConfig: JiraConfig): Promise<Task[]>;
}

export interface JiraClient {
  getData(url: string): Promise<any>;
}

export interface JiraRepository {
  issueDetailsWithChangelog(
    jiraConfig: JiraConfig,
    issueId: string
  ): Promise<IssueDetails>;
  sprintData(jiraConfig: JiraConfig, sprint: Sprint): Promise<Task[]>;

  completedSprints(
    teamId: number,
    since: Date,
    until?: Date
  ): Promise<Sprint[]>;

  completedKanbanIssuesSince(jiraConfig: JiraConfig): Promise<Task[]>;
}
