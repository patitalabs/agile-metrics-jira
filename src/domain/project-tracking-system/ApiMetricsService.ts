import { Logger } from '../../config/Logger';
import {
  PtsCollectorService,
  CoreMetricsService,
  TeamMetricsRequest,
} from '../Types';

export class ApiMetricsService {
  constructor(
    private readonly coreMetricsService: CoreMetricsService,
    private readonly ciCollectorService: PtsCollectorService
  ) {}

  public async metricsForRequest(
    teamMetricsRequest: TeamMetricsRequest
  ): Promise<void> {
    try {
      Logger.info(`teamMetricsRequest: ${JSON.stringify(teamMetricsRequest)}`);
      const configurationDescriptors =
        ApiMetricsService.createConfigurationDescriptorsForRequest(
          teamMetricsRequest
        );
      Logger.info(
        `created configurations: ${JSON.stringify(configurationDescriptors)}`
      );

      for (const configurationDescriptor of configurationDescriptors) {
        const scmMetricItems = await this.ciCollectorService.fetch(
          configurationDescriptor.config
        );
        await this.coreMetricsService.publish(
          scmMetricItems,
          configurationDescriptor.shouldUpdateEntries
        );
      }
      Logger.info('Done!');
    } catch (error) {
      Logger.error(error);
    }
  }

  static createConfigurationDescriptorsForRequest(
    teamMetricsRequest: TeamMetricsRequest
  ): any[] {
    if (teamMetricsRequest.config) {
      return [{ ...teamMetricsRequest }];
    }

    throw Error(
      `Unable to create configuration for request ${JSON.stringify(
        teamMetricsRequest
      )}`
    );
  }
}
