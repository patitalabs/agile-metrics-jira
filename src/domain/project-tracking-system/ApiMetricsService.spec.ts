import { ApiMetricsService } from './ApiMetricsService';
import { TeamMetricsRequest } from '../Types';

describe('ApiMetricsService', () => {
  it('should fail with invalid configuration', () => {
    const teamMetricsRequest: TeamMetricsRequest = {
      shouldUpdateEntries: false,
      config: { fake: 'config' },
    };
    const referenceDate = new Date('2019-02-09');

    try {
      ApiMetricsService.createConfigurationDescriptorsForRequest(
        teamMetricsRequest
      );
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it('should return existing unchanged config', () => {
    const teamMetricsRequest: TeamMetricsRequest = {
      shouldUpdateEntries: false,
      config: { fake: 'config' },
    };
    const referenceDate = new Date('2019-02-12');
    const generatedConfigs =
      ApiMetricsService.createConfigurationDescriptorsForRequest(
        teamMetricsRequest
      );
    expect(generatedConfigs).toMatchSnapshot();
  });
});
