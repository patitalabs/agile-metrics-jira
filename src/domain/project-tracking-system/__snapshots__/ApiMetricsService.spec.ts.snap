// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiMetricsService should return existing unchanged config 1`] = `
Array [
  Object {
    "config": Object {
      "fake": "config",
    },
    "shouldUpdateEntries": false,
  },
]
`;
